docs/proposals/distributed-query-execution.md

```diff
@@ -0,0 +1,143 @@
+
+![LogicalPlan](/images/proposals/distributed-execution-logical-plan.png)
+
+#### Distributed Optimizer
+
+A special optimizer, the **Distributed Optimizer**, analyzes the logical plan to determine where and how the query can be split and executed in parallel. This is done by inserting **`Remote` logical nodes** at appropriate points in the plan tree.
+These `Remote` nodes act as placeholders indicating plan fragments that can be offloaded to remote queriers. This enables sharding both over time and across data partitions.
+
+![OptimizedLogicalPlan](/images/proposals/distributed-execution-optimized-plan.png)
+
+### Query Scheduler
+
+The **Query Scheduler** receives the optimized logical plan, including the `Remote` nodes introduced by the optimizer. It is responsible for **fragmenting** the plan and orchestrating its execution across multiple queriers.
+
+#### Fragmentation
+
+Fragmentation involves splitting the query plan into **multiple fragments**, where each fragment is a sub-tree of the overall plan that can be evaluated independently. Fragments are created by cutting the plan at `Remote` nodes. In the below diagram the coloured tiles mark the same fragment.
+
+![PlanFragments](/images/proposals/distributed-execution-plan-fragments.png)
+
+#### Enqueuing fragments
+
+Fragmentation respects dependencies: **child fragments are enqueued before parent fragments**, ensuring data is available before dependent operations are scheduled.
+These fragments are then **enqueued** into the scheduler’s internal queue system, ready to be picked up by queriers.
+
+![SchedulerQueue](/images/proposals/distributed-execution-scheduler-queue.png)
+
+
+#### Assigning query fragments to queriers
+
+Each fragment is assigned to a querier based on resource availability, tenant isolation, and data locality (if applicable). The fragment include the plan, the remote fragment IDs it depends on. After child fragments are assigned to a particular querier, it’s parent fragment is updated with the location of the querier from which to read the results from.
+
+At execution time, a `Remote` operator acts as a placeholder within the querier's execution engine to **pull results from other queriers** running dependent fragments.
+
+![AssignedFragments](/images/proposals/distributed-execution-assigned-fragments.png)
+
+### Querier
+
+The **Querier** is enhanced to support both **executing** fragments and **serving results** to other queriers. It now implements two interfaces:
+
+* **Executor** – Responsible for executing the root fragment (or any assigned fragment) of the physical query plan.
+* **Server** – Serves results for executed fragments to other queriers that depend on them (i.e., acts as a remote data source for `Remote` operators).
```

Comment:
Maybe we can add more details about the two interfaces? Like what methods are required.

To me `Executor` seems not an interface but something already can be done by the query API + support plan/fragment as a parameter.

`Server` interface is probably similar to the operator?

```
type Server interface {
  Series() []Series
  Next() []StepVector
}
```

Reply:
I added an example interface for reference. It's not rigid and can be changed during the actual implementation.